### PR TITLE
Batch submit to support setting python 2 or 3

### PR DESCRIPTION
**Pull request by KevinGrealish**
Add ability to set Python version for batch submit. See: [https://issues.cloudera.org/browse/LIVY-159](https://issues.cloudera.org/browse/LIVY-159)

Added new fields to batch submit that echo the environment variables that are accepted by spark-submit; PYSPARK_PYTHON and PYSPARK_DRIVER_PYTHON. To prevent ability to execute arbitrary code via these, only "python2" and "python3" are permitted.

### CODE REVIEW

**Comment by vanzin** [comment]
Hi @KevinGrealish, we need you to sign an ICLA before we can merge patches from you. MS has already signed a CCLA so you're good on that front.

You can find the info on the wiki: https://github.com/cloudera/livy/wiki/Contributing-to-Livy

**Comment by KevinGrealish** [comment]
I thought the MS one covered me. Are you saying you need both?

**Comment by vanzin** [comment]
Let me ask around. I was under the impression or legal people required both. :-/

File: server/src/main/scala/com/cloudera/livy/server/batch/BatchSession.scala [hunk 1]
```diff
@@ -38,6 +38,10 @@ class BatchSession(
     val conf = prepareConf(request.conf, request.jars, request.files, request.archives,
       request.pyFiles)
     require(request.file != null, "File is required.")
+    require(!request.pyspark_python.exists(s => s != "python2" && s != "python3"),
```
Comment (felixcheung) [inline server/src/main/scala/com/cloudera/livy/server/batch/BatchSession.scala]: I think in a lot of environment, Python 2 is simply `python` (eg. CentOS) - could this support that?

**Comment by ksakellis** [comment]
@KevinGrealish You will need to get both the ICLA:
https://github.com/cloudera/livy/wiki/Individual-Contributor-License-Agreement-(ICLA)
and get another CCLA:
https://github.com/cloudera/livy/wiki/Corporate-Contributor-License-Agreement-(CCLA)
signed by your employer Microsoft with your name under the contributors list. 

Each contributor needs to have their name listed on the CCLA from their employer and have an ICLA signed. Please let me know if you have any questions.

**Comment by sairayapureddy** [comment]
Can any one suggest me a documentation or resource for following requirement.
I need to invoke python scripts present in spark from node js server (using livy restful API). I need to check the status of submitted job. This seems basic but I am a starter and started exploring from 2 days.

**Comment by Stibbons** [comment]
Hello. I am not part of the project but very interested in it. Just to give my feedback on this pull request

For the moment we cannot define easily the python environment on Spark, so it might be pretty common to have let's say N environment preinstalled on the Spark cluster's executor. For example, one "Python 2 with pandas" environment, one "Python 3 with tensorflow", one "Anaconda Python 2 with pretty much all libraries", etc... Each are installed on the same folder  on each executor.

To deploy our spark job, we need a way to make them point to the right environment, using the PYSPARK_PYTHON environment variable for instance. If you limit the possible values, your restrict the use of the system python and prevent usage of preinstalled virtual environments.

Let me know what you think.
Regards

**Comment by KevinGrealish** [comment]
That was one of the reasons this change was not made and instead the Spark config was fixed instead so that the python environment variables can be set to any value and Spark will respect them (the conf values were ignored in YARN cluster mode if they were already set system environment vars). See:
https://issues.apache.org/jira/browse/SPARK-16110
https://github.com/apache/spark/pull/13824/files

**Pull request closed without merging**